File: Labs/Lab-04-20231013/app/src/main/java/sap/escooters/domain_layer/EScooter.java
package sap.escooters.domain_layer;

import java.util.Optional;

import io.vertx.core.json.JsonObject;

public class EScooter  {

	private String id;
	public enum EScooterState { AVAILABLE, IN_USE, MAINTENANCE}	
	private EScooterState state;
	private Optional<Location> loc;
	
	public EScooter(String id) {
		this.id = id;
		this.state = EScooterState.AVAILABLE;
		this.loc = Optional.empty();
	}
	
	public String getId() {
		return id;
	}

	public EScooterState getState() {
		return state;
	}
	
	public boolean isAvailable() {
		return state.equals(EScooterState.AVAILABLE);
	}

	public void updateState(EScooterState state) {
		this.state = state;
		save();
	}
	
	public void updateLocation(Location newLoc) {
		loc = Optional.of(newLoc);
		save();
	}
	
	public Optional<Location> getCurrentLocation(){
		return loc;
	}
	
	public void save() {
		try {
			DomainLayerImpl.getDataSourceLayer().saveEScooter(toJson());
		} catch (Exception ex) {
			ex.printStackTrace();
		}
	}
	
	public JsonObject toJson() {
		JsonObject scooterObj = new JsonObject();
		scooterObj.put("id", this.getId());
		scooterObj.put("state", this.getState().toString());
		Optional<Location> loc = this.getCurrentLocation();
		if (loc.isPresent()) {
			JsonObject locObj = new JsonObject();
			locObj.put("latitude", this.getCurrentLocation().get().getLatitude());
			locObj.put("longitude", this.getCurrentLocation().get().getLongitude());
			scooterObj.put("location", locObj);			
		} else {
			scooterObj.putNull("location");			
		}			
		return scooterObj;
	}

	
}

File: Labs/Lab-05-20231016/app/src/main/kotlin/assigment03/escooters/domain_layer/EScooter.kt
package assigment03.escooters.domain_layer

import io.vertx.core.json.JsonObject

class EScooter(val id: String) {
    enum class EScooterState { AVAILABLE, IN_USE, MAINTENANCE }
    var state = EScooterState.AVAILABLE
        private set
    private var loc: Location? = null

    fun isAvailable() = state == EScooterState.AVAILABLE
    fun updateState(state: EScooterState) {
        this.state = state
        save()
    }

    fun updateLocation(newLoc: Location) {
        loc = newLoc
        save()
    }

    fun getCurrentLocation() = loc
    fun save() {
        try {
            DomainLayerImpl.getDataSourceLayer()?.saveEScooter(toJson())
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    fun toJson() = JsonObject().apply {
        put("id", id)
        put("state", state.toString())
        getCurrentLocation()?.let {
            put("location", JsonObject().put("latitude", it.latitude).put("longitude", it.longitude))
        } ?: putNull("location")
    }
}